# IBM Applied Data Science Capstone Course by Coursera

**_Opening a New Shopping Mall in New Delhi, India_**
- Build a dataframe of neighborhoods in New Delhi, India by scraping the data from Wikipedia page
- Get the geographical coordinates of the neighborhoods
- Obtain the venue data for the neighborhoods from Foursquare API
- Explore and cluster the neighborhoods
- Select the best cluster to open a new shopping mall
***
### 1. Import libraries

In [7]:
import numpy as np

import pandas as pd
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values # to get coordinates

import requests
import geocoder # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print("Libraries  are imported.")

Libraries  are imported.


### 2. Scrap data from Wikipedia page into a DataFrame

In [8]:
# send the GET request
data = requests.get("https://en.wikipedia.org/wiki/Category:Neighbourhoods_in_Delhi").text

In [9]:
# parse data from the html into a beautifulsoup object
soup = BeautifulSoup(data, 'html.parser')

In [10]:
# create a list to store neighborhood data
neighborhoodList = []

In [11]:
# append the data into the list
for row in soup.find_all("div", class_="mw-category")[0].findAll("li"):
    neighborhoodList.append(row.text)
neighborhoodList.remove('Neighbourhoods of Delhi')

In [12]:
# create a new DataFrame from the list
kl_df = pd.DataFrame({"Neighborhood": neighborhoodList})
kl_df.head()

Neighborhood
0  Ashok Nagar (Delhi)
1          Ashok Vihar
2         Ashram Chowk
3             Babarpur
4      Badarpur, Delhi

In [13]:
kl_df.shape

(141, 1)

### 3. Get the geographical coordinates

In [14]:
# define a function to get coordinates
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, New Delhi, India'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [15]:
# call the function to get the coordinates, store in a new list using list comprehension
coords = [ get_latlng(neighborhood) for neighborhood in kl_df["Neighborhood"].tolist() ]

In [16]:
coords

[[28.692230000000052, 77.30124000000006],
 [28.69037000000003, 77.17609000000004],
 [28.710597501792023, 77.32696517369723],
 [28.50738000000007, 77.30346000000003],
 [28.50738000000007, 77.30346000000003],
 [28.652234222889238, 77.12939224396462],
 [28.79767000000004, 77.04522000000003],
 [28.549540000000036, 77.18167000000005],
 [28.699880000000064, 77.25906000000003],
 [28.595060000000046, 77.18573000000004],
 [28.656270000000063, 77.23232000000007],
 [28.538400000000024, 77.24832000000004],
 [28.67671000000007, 77.21767000000006],
 [28.633940000000052, 77.21968000000004],
 [28.60761000000008, 77.08714000000003],
 [28.65457890544559, 77.23339989939495],
 [28.62832000000003, 77.24727000000007],
 [28.605920000000026, 77.08529000000004],
 [28.560590000000047, 77.24678000000006],
 [28.57298000000003, 77.23357000000004],
 [28.591510000000028, 77.12945000000008],
 [28.699110000000076, 77.19105000000008],
 [28.594857177133914, 77.16729160908383],
 [28.684700000000078, 77.32774000000006],
 

In [17]:
# create temporary dataframe to populate the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [18]:
# merge the coordinates into the original dataframe
kl_df['Latitude'] = df_coords['Latitude']
kl_df['Longitude'] = df_coords['Longitude']

In [19]:
# check the neighborhoods and the coordinates
print(kl_df.shape)
kl_df

(141, 3)


Neighborhood   Latitude  Longitude
0                            Ashok Nagar (Delhi)  28.692230  77.301240
1                                    Ashok Vihar  28.690370  77.176090
2                                   Ashram Chowk  28.710598  77.326965
3                                       Babarpur  28.507380  77.303460
4                                Badarpur, Delhi  28.507380  77.303460
5                                     Bali Nagar  28.652234  77.129392
6                                         Bawana  28.797670  77.045220
7                                      Ber Sarai  28.549540  77.181670
8                                     Bhajanpura  28.699880  77.259060
9                                   Chanakyapuri  28.595060  77.185730
10                                 Chandni Chowk  28.656270  77.232320
11                             Chittaranjan Park  28.538400  77.248320
12                            Civil Lines, Delhi  28.676710  77.217670
13                    Connaught Place, New Delhi  28.633940  77.219680
14                              Dabri, New Delhi  28.607610  77.087140
15                                  Dariba Kalan  28.654579  77.233400
16                                     Daryaganj  28.628320  77.247270
17                                 Dashrath Puri  28.605920  77.085290
18                               Dayanand Colony  28.560590  77.246780
19                                Defence Colony  28.572980  77.233570
20                              Delhi Cantonment  28.591510  77.129450
21                                 Derawal Nagar  28.699110  77.191050
22                                   Dhaula Kuan  28.594857  77.167292
23                                Dilshad Colony  28.684700  77.327740
24                                Dilshad Garden  28.679040  77.314760
25                                 Dwarka, Delhi  28.589950  77.040040
26                              East Patel Nagar  28.648170  77.178330
27                 Gadaipur, Mehrauli, New Delhi  28.479270  77.150540
28                           Gandhi Nagar, Delhi  28.724530  77.207680
29                            Geetanjali Enclave  28.528450  77.203540
30                                      Ghitorni  28.498300  77.135960
31                                   Gole Market  28.634100  77.205690
32                         Golf Links, New Delhi  28.603040  77.232690
33                                    Govindpuri  28.535080  77.265120
34                               Greater Kailash  28.548490  77.236670
35                             Green Park, Delhi  28.558970  77.204620
36                                   Gulabi Bagh  28.620430  77.049410
37                                 Gulmohar Park  28.554390  77.212520
38                                     Hauz Khas  28.551090  77.203990
39                                    Inder Puri  28.628030  77.145040
40                                     Janakpuri  28.627910  77.090600
41                                      Jangpura  28.583400  77.247190
42                                     Jia Sarai  28.546380  77.188970
43                        Kabir Nagar, New Delhi  28.689664  77.141038
44                                Kailash Colony  28.556130  77.240600
45                        Kamla Nagar, New Delhi  28.681628  77.201826
46                       Kapasheda Border, Delhi  28.521630  77.086450
47                                    Karol Bagh  28.650450  77.188730
48                                  Keshav Puram  28.688010  77.158660
49                                 Khaira, Delhi  28.594100  76.970360
50                                Khanpur, Delhi  28.509630  77.231080
51                                   Khari Baoli  28.657260  77.222840
52                                 Kingsway Camp  28.711690  77.201970
53                                   Kirti Nagar  28.648210  77.142730
54                      Kotla Mubarakpur Complex  28.574350  77.224190
55                          Krishna Nagar, Delhi  28.563640 

In [20]:
# save the DataFrame as CSV file
kl_df.to_csv("kl_df.csv", index=False)

### 4. Create a map of New Delhi with neighborhoods superimposed on top

In [21]:
# get the coordinates of New Delhi
address = 'New Delhi, India'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New Delhi, India {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New Delhi, India 28.6141793, 77.2022662.


In [22]:
# create map of New Delhi using latitude and longitude values
map_kl = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, neighborhood in zip(kl_df['Latitude'], kl_df['Longitude'], kl_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_kl)  
    
map_kl

In [24]:
# save the map as HTML file
map_kl.save('map_kl.html')

### 5. Use the Foursquare API to explore the neighborhoods

In [1]:
# define Foursquare Credentials and Version
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20200202' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 
CLIENT_SECRET:


**Now, let's get the top 100 venues that are within a radius of 2000 meters.**

In [27]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(kl_df['Latitude'], kl_df['Longitude'], kl_df['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [28]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(6201, 7)


Neighborhood  Latitude  Longitude                      VenueName  \
0  Ashok Nagar (Delhi)  28.69223   77.30124                    Sutta Chowk   
1  Ashok Nagar (Delhi)  28.69223   77.30124                    AFM PVT LTD   
2  Ashok Nagar (Delhi)  28.69223   77.30124                   yamuna vihar   
3  Ashok Nagar (Delhi)  28.69223   77.30124                   Shivaji park   
4  Ashok Nagar (Delhi)  28.69223   77.30124  Mansarover Park Metro Station   

   VenueLatitude  VenueLongitude               VenueCategory  
0      28.697897       77.300010                  Smoke Shop  
1      28.704770       77.309608  Tourist Information Center  
2      28.689816       77.283876                        Park  
3      28.682657       77.285503                        Park  
4      28.675370       77.300932               Train Station

**Let's check how many venues were returned for each neighorhood**

In [29]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  \
Neighborhood                                                                  
Ashok Nagar (Delhi)                                 5          5          5   
Ashok Vihar                                        22         22         22   
Ashram Chowk                                        5          5          5   
Babarpur                                            4          4          4   
Badarpur, Delhi                                     4          4          4   
Bali Nagar                                         51         51         51   
Bawana                                              2          2          2   
Ber Sarai                                          91         91         91   
Bhajanpura                                          5          5          5   
Chanakyapuri                                       74         74         74   
Chandni Chowk                                      53         53         53   
Chittaranjan Park                                  91         91         91   
Civil Lines, Delhi                                 40         40         40   
Connaught Place, New Delhi                        100        100        100   
Dabri, New Delhi                                    4          4          4   
Dariba Kalan                                       53         53         53   
Daryaganj                                          40         40         40   
Dashrath Puri                                       6          6          6   
Dayanand Colony                                   100        100        100   
Defence Colony                                     95         95         95   
Delhi Cantonment                                    8          8          8   
Derawal Nagar                                      32         32         32   
Dhaula Kuan                                        41         41         41   
Dilshad Colony                                      7          7          7   
Dilshad Garden                                     11         11         11   
Dwarka, Delhi                                      31         31         31   
East Patel Nagar                                   51         51         51   
Gadaipur, Mehrauli, New Delhi                      12         12         12   
Gandhi Nagar, Delhi                                 8          8          8   
Geetanjali Enclave                                100        100        100   
Ghitorni                                            5          5          5   
Gole Market                                       100        100        100   
Golf Links, New Delhi                              95         95         95   
Govindpuri                                         26         26         26   
Greater Kailash                                   100        100        100   
Green Park, Delhi                                 100        100        100   
Gulabi Bagh                                        10         10         10   
Gulmohar Park                                     100        100        100   
Hauz Khas                                         100        100        100   
Inder Puri                                          9          9          9   
Janakpuri                                          27         27         27   
Jangpura                                           80         80         80   
Jia Sarai                                          95         95         95   
Kabir Nagar, New Delhi                             27         27         27   
Kailash Colony                                    100        100        100   
Kamla Nagar, New Delhi                             37         37         37   
Kapasheda Border, Delhi                            19         19         19   
Karol Bagh                                         46         46         46   
Keshav Puram                                       32         32         32   
Khaira, Delhi   

**Let's find out how many unique categories can be curated from all the returned venues**

In [30]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 213 uniques categories.


In [31]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:50]

array(['Smoke Shop', 'Tourist Information Center', 'Park',
       'Train Station', 'Athletics & Sports', 'Asian Restaurant',
       'Sandwich Place', 'Snack Place', 'Pizza Place',
       'Indian Restaurant', 'South Indian Restaurant', 'Department Store',
       'Fast Food Restaurant', 'Coffee Shop', 'Market', 'Dessert Shop',
       'Basketball Court', 'Light Rail Station', 'ATM', 'Print Shop',
       'Indian Sweet Shop', 'Café', 'American Restaurant', 'Diner',
       'Donut Shop', 'Bakery', 'Hookah Bar', 'BBQ Joint', 'Sports Bar',
       'Hotel', 'Garden Center', 'Pub', 'Fried Chicken Joint',
       'Shopping Mall', 'Furniture / Home Store', 'Multiplex',
       'Metro Station', 'Bar', 'Gym', 'Gym / Fitness Center', 'Garden',
       'Convenience Store', 'Salon / Barbershop', 'Playground',
       'Art Gallery', 'Mediterranean Restaurant', 'Tibetan Restaurant',
       'Tea Room', 'Lounge', 'Ice Cream Shop'], dtype=object)

In [32]:
# check if the results contain "Shopping Mall"
"Neighborhood" in venues_df['VenueCategory'].unique()

True

### 6. Analyze Neighborhood

In [33]:
# one hot encoding
kl_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
kl_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [kl_onehot.columns[-1]] + list(kl_onehot.columns[:-1])
kl_onehot = kl_onehot[fixed_columns]

print(kl_onehot.shape)
kl_onehot.head()

(6201, 214)


Neighborhoods  ATM  Accessories Store  Airport  Airport Food Court  \
0  Ashok Nagar (Delhi)    0                  0        0                   0   
1  Ashok Nagar (Delhi)    0                  0        0                   0   
2  Ashok Nagar (Delhi)    0                  0        0                   0   
3  Ashok Nagar (Delhi)    0                  0        0                   0   
4  Ashok Nagar (Delhi)    0                  0        0                   0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Arcade  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0            0           0                    0                 0   
1       0            0           0                    0                 0   
2       0            0           0                    0                 0   
3       0            0           0                    0                 0   
4       0            0           0                    0                 0   

   Astrologer  Athletics & Sports  Australian Restaurant  BBQ Joint  \
0           0                   0                      0          0   
1           0                   0                      0          0   
2           0                   0                      0          0   
3           0                   0                      0          0   
4           0                   0                      0          0   

   Bagel Shop  Bakery  Bank  Bar  Basketball Court  Bed & Breakfast  \
0           0       0     0    0                 0                0   
1           0       0     0    0                 0                0   
2           0       0     0    0                 0                0   
3           0       0     0    0                 0                0   
4           0       0     0    0                 0                0   

   Beer Garden  Bengali Restaurant  Big Box Store  Bike Shop  Bistro  \
0            0                   0              0          0       0   
1            0                   0              0          0       0   
2            0                   0              0          0       0   
3            0                   0              0          0       0   
4            0                   0              0          0       0   

   Boat or Ferry  Bookstore  Boutique  Breakfast Spot  Building  Burger Joint  \
0              0          0         0               0         0             0   
1              0          0         0               0         0             0   
2              0          0         0               0         0             0   
3              0          0         0               0         0             0   
4              0          0         0               0         0             0   

   Burmese Restaurant  Bus Station  Business Service  Cafeteria  Café  \
0                   0            0                 0          0     0   
1                   0            0                 0          0     0   
2                   0            0                 0          0     0   
3                   0            0                 0          0     0   
4                   0            0                 0          0     0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  \
0                   0               0             0            0   
1                   0               0             0            0   
2                   0               0             0            0   
3                   0               0             0            0   
4                   0               0             0            0   

**Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category**

In [34]:
kl_grouped = kl_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(kl_grouped.shape)
kl_grouped

(141, 214)


Neighborhoods       ATM  Accessories Store  \
0                            Ashok Nagar (Delhi)  0.000000           0.000000   
1                                    Ashok Vihar  0.000000           0.000000   
2                                   Ashram Chowk  0.800000           0.000000   
3                                       Babarpur  0.000000           0.000000   
4                                Badarpur, Delhi  0.000000           0.000000   
5                                     Bali Nagar  0.000000           0.000000   
6                                         Bawana  0.500000           0.000000   
7                                      Ber Sarai  0.000000           0.000000   
8                                     Bhajanpura  0.600000           0.000000   
9                                   Chanakyapuri  0.000000           0.000000   
10                                 Chandni Chowk  0.000000           0.000000   
11                             Chittaranjan Park  0.000000           0.000000   
12                            Civil Lines, Delhi  0.000000           0.000000   
13                    Connaught Place, New Delhi  0.000000           0.000000   
14                              Dabri, New Delhi  0.000000           0.000000   
15                                  Dariba Kalan  0.000000           0.000000   
16                                     Daryaganj  0.000000           0.000000   
17                                 Dashrath Puri  0.000000           0.000000   
18                               Dayanand Colony  0.000000           0.000000   
19                                Defence Colony  0.000000           0.000000   
20                              Delhi Cantonment  0.000000           0.000000   
21                                 Derawal Nagar  0.000000           0.000000   
22                                   Dhaula Kuan  0.000000           0.000000   
23                                Dilshad Colony  0.000000           0.000000   
24                                Dilshad Garden  0.000000           0.000000   
25                                 Dwarka, Delhi  0.000000           0.000000   
26                              East Patel Nagar  0.000000           0.000000   
27                 Gadaipur, Mehrauli, New Delhi  0.000000           0.000000   
28                           Gandhi Nagar, Delhi  0.250000           0.125000   
29                            Geetanjali Enclave  0.000000           0.000000   
30                                      Ghitorni  0.000000           0.000000   
31                                   Gole Market  0.000000           0.000000   
32                         Golf Links, New Delhi  0.000000           0.000000   
33                                    Govindpuri  0.000000           0.000000   
34                               Greater Kailash  0.000000           0.000000   
35                             Green Park, Delhi  0.000000           0.000000   
36                                   Gulabi Bagh  0.000000           0.000000   
37                                 Gulmohar Park  0.000000           0.000000   
38                                     Hauz Khas  0.000000           0.000000   
39                                    Inder Puri  0.000000           0.000000   
40                                     Janakpuri  0.000000           0.000000   
41                                      Jangpura  0.000000           0.000000   
42                                     Jia Sarai  0.000000           0.000000   
43                        Kabir Nagar, New Delhi  0.000000           0.000000   
44                                Kailash Colony  0.000000           0.000000   
45                        Kamla Nagar, New Delhi  0.000000           0.000000   
46                       Kapasheda Border, Delhi  0.000000           0.000000   
47                                    Karol Bagh  0.000000           0.000000   
48                                  Keshav Puram  0.000000        

In [35]:
len(kl_grouped[kl_grouped["Shopping Mall"] > 0])

36

**Create a new DataFrame for Shopping Mall data only**

In [36]:
kl_mall = kl_grouped[["Neighborhoods","Shopping Mall"]]

In [37]:
kl_mall.head()

Neighborhoods  Shopping Mall
0  Ashok Nagar (Delhi)            0.0
1          Ashok Vihar            0.0
2         Ashram Chowk            0.0
3             Babarpur            0.0
4      Badarpur, Delhi            0.0

### 7. Cluster Neighborhoods
Run k-means to cluster the neighborhoods in New Delhi into 3 clusters.

In [38]:
# set number of clusters
kclusters = 3

kl_clustering = kl_mall.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(kl_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 2, 0, 0, 0, 0], dtype=int32)

In [39]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
kl_merged = kl_mall.copy()

# add clustering labels
kl_merged["Cluster Labels"] = kmeans.labels_

In [40]:
kl_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
kl_merged.head()

Neighborhood  Shopping Mall  Cluster Labels
0  Ashok Nagar (Delhi)            0.0               0
1          Ashok Vihar            0.0               0
2         Ashram Chowk            0.0               0
3             Babarpur            0.0               0
4      Badarpur, Delhi            0.0               0

In [41]:
# merge dataFrames to add latitude/longitude for each neighborhood
kl_merged = kl_merged.join(kl_df.set_index("Neighborhood"), on="Neighborhood")

print(kl_merged.shape)
kl_merged.head() # check the last columns!

(141, 5)


Neighborhood  Shopping Mall  Cluster Labels   Latitude  Longitude
0  Ashok Nagar (Delhi)            0.0               0  28.692230  77.301240
1          Ashok Vihar            0.0               0  28.690370  77.176090
2         Ashram Chowk            0.0               0  28.710598  77.326965
3             Babarpur            0.0               0  28.507380  77.303460
4      Badarpur, Delhi            0.0               0  28.507380  77.303460

In [42]:
# sort the results by Cluster Labels
print(kl_merged.shape)
kl_merged.sort_values(["Cluster Labels"], inplace=True)
kl_merged

(141, 5)


Neighborhood  Shopping Mall  \
0                            Ashok Nagar (Delhi)       0.000000   
92                                     Paharganj       0.000000   
91                                     Old Delhi       0.000000   
90                                         Okhla       0.000000   
89                               Nizamuddin West       0.000000   
88                               Nizamuddin East       0.000000   
87                                Nigambodh Ghat       0.000000   
86                                 New Moti Bagh       0.011628   
85                            New Friends Colony       0.000000   
84                                     New Delhi       0.000000   
83                           Netaji Nagar, Delhi       0.000000   
82                               Naveen Shahdara       0.000000   
81                                        Narela       0.000000   
80                                  Naraina Area       0.000000   
79                                   Nangloi Jat       0.000000   
78                                    Nand Nagri       0.000000   
77                                     Nanakpura       0.000000   
76                                     Najafgarh       0.000000   
75                                       Munirka       0.019231   
74                               Mukherjee Nagar       0.000000   
72                                     Moti Bagh       0.020000   
71                            Model Town (Delhi)       0.000000   
139                                Yamuna Pushta       0.000000   
69                                    Meera Bagh       0.000000   
68                         Mayur Vihar Phase - 3       0.000000   
66                                      Mayapuri       0.000000   
65                         Malviya Nagar (Delhi)       0.020000   
64                                Majnu-ka-tilla       0.000000   
93                                         Palam       0.000000   
63                                    Mahipalpur       0.015152   
94                                 Palika Bazaar       0.000000   
97                                 Paschim Vihar       0.000000   
138                             West Patel Nagar       0.000000   
136                           Vasundhara Enclave       0.000000   
135                          Vasant Vihar, Delhi       0.018519   
133                                  Urdu Bazaar       0.000000   
132                                   Tis Hazari       0.000000   
130                                 Sriniwaspuri       0.000000   
129                            South Patel Nagar       0.000000   
128                              South Extension       0.000000   
126                                Shiv Ram Park       0.000000   
125                                 Shastri Park       0.000000   
124                        Shanti Niketan, Delhi       0.018868   
123                                Shankar Vihar       0.000000   
121                          Shakti Nagar, Delhi       0.000000   
120                                  Shahpur Jat       0.000000   
119                            Shahdara district       0.000000   
118                        Seemapuri subdivision       0.000000   
117                        Seelampur subdivision       0.000000   
115                                 Sarita Vihar       0.000000   
114                              Sarai Kale Khan       0.000000   
111                                    Sagar Pur       0.000000   
110                           Safdarjung (Delhi)       0.000000   
108                                   Roop Nagar       0.000000   
105                           Rama Krishna Puram       0.013514   
103                               Rajendra Place       0.000000   
102                        Rajendra Nagar, Delhi       0.000000   
101                                 Raisina Hill       0.000000   
98                                   Patel Nagar       0.000000   
95              

**Finally, let's visualize the resulting clusters**

In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(kl_merged['Latitude'], kl_merged['Longitude'], kl_merged['Neighborhood'], kl_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [44]:
# save the map as HTML file
map_clusters.save('map_clusters.html')

### 8. Examine Clusters

#### Cluster 0

In [45]:
kl_merged.loc[kl_merged['Cluster Labels'] == 0].shape

(116, 5)

In [46]:
kl_merged.loc[kl_merged['Cluster Labels'] == 0]

Neighborhood  Shopping Mall  Cluster Labels   Latitude  \
0              Ashok Nagar (Delhi)       0.000000               0  28.692230   
92                       Paharganj       0.000000               0  28.645960   
91                       Old Delhi       0.000000               0  28.654340   
90                           Okhla       0.000000               0  28.532470   
89                 Nizamuddin West       0.000000               0  28.588650   
88                 Nizamuddin East       0.000000               0  28.602462   
87                  Nigambodh Ghat       0.000000               0  28.664710   
86                   New Moti Bagh       0.011628               0  28.580997   
85              New Friends Colony       0.000000               0  28.578120   
84                       New Delhi       0.000000               0  28.630950   
83             Netaji Nagar, Delhi       0.000000               0  28.577460   
82                 Naveen Shahdara       0.000000               0  28.673690   
81                          Narela       0.000000               0  28.839790   
80                    Naraina Area       0.000000               0  28.633660   
79                     Nangloi Jat       0.000000               0  28.678650   
78                      Nand Nagri       0.000000               0  28.696710   
77                       Nanakpura       0.000000               0  28.622880   
76                       Najafgarh       0.000000               0  28.625100   
75                         Munirka       0.019231               0  28.555040   
74                 Mukherjee Nagar       0.000000               0  28.710530   
72                       Moti Bagh       0.020000               0  28.583620   
71              Model Town (Delhi)       0.000000               0  28.705010   
139                  Yamuna Pushta       0.000000               0  28.621180   
69                      Meera Bagh       0.000000               0  28.661210   
68           Mayur Vihar Phase - 3       0.000000               0  28.611250   
66                        Mayapuri       0.000000               0  28.623340   
65           Malviya Nagar (Delhi)       0.020000               0  28.533940   
64                  Majnu-ka-tilla       0.000000               0  28.698930   
93                           Palam       0.000000               0  28.591060   
63                      Mahipalpur       0.015152               0  28.548420   
94                   Palika Bazaar       0.000000               0  28.631560   
97                   Paschim Vihar       0.000000               0  28.669330   
138               West Patel Nagar       0.000000               0  28.647830   
136             Vasundhara Enclave       0.000000               0  28.600150   
135            Vasant Vihar, Delhi       0.018519               0  28.564940   
133                    Urdu Bazaar       0.000000               0  28.648881   
132                     Tis Hazari       0.000000               0  28.666000   
130                   Sriniwaspuri       0.000000               0  28.565680   
129              South Patel Nagar       0.000000               0  28.645610   
128                South Extension       0.000000               0  28.562490   
126                  Shiv Ram Park       0.000000               0  28.664450   
125                   Shastri Park       0.000000               0  28.670060   
124          Shanti Niketan, Delhi       0.018868               0  28.577560   
123                  Shankar Vihar       0.000000               0  28.566400   
121            Shakti Nagar, Delhi       0.000000               0  28.675570   
120                    Shahpur Jat       0.000000               0  28.548540   
119              Shahdara district       0.000000               0  28.689570   
118          Seemapuri subdivision       0.000000               0  28.687600   
117          Seelampur subdivision       0.000000               0  28.689440   
115                   

#### Cluster 1

In [47]:
kl_merged.loc[kl_merged['Cluster Labels'] == 1].shape

(7, 5)

In [48]:
kl_merged.loc[kl_merged['Cluster Labels'] == 1]

Neighborhood  Shopping Mall  Cluster Labels  Latitude  \
57       Laxmi Nagar (Delhi)       0.105263               1  28.63875   
46   Kapasheda Border, Delhi       0.105263               1  28.52163   
20          Delhi Cantonment       0.125000               1  28.59151   
48              Keshav Puram       0.093750               1  28.68801   
107            Rohini, Delhi       0.100000               1  28.73356   
99                Pitam Pura       0.111111               1  28.69590   
109      Sadar Bazaar, Delhi       0.125000               1  28.59028   

     Longitude  
57    77.27592  
46    77.08645  
20    77.12945  
48    77.15866  
107   77.10401  
99    77.13725  
109   77.12014

#### Cluster 2

In [49]:
kl_merged.loc[kl_merged['Cluster Labels'] == 2].shape

(18, 5)

In [50]:
kl_merged.loc[kl_merged['Cluster Labels'] == 2]

Neighborhood  Shopping Mall  \
137                      Vivek Vihar subdivision       0.060606   
134                                  Vasant Kunj       0.048387   
96                                  Pandav Nagar       0.058824   
5                                     Bali Nagar       0.039216   
131                          Tilak Nagar (Delhi)       0.043478   
67                                   Mayur Vihar       0.058824   
100                                 Punjabi Bagh       0.048780   
53                                   Kirti Nagar       0.058824   
73                        Moti Nagar (New Delhi)       0.055556   
122  Shalimar Bagh (Delhi Assembly constituency)       0.066667   
25                                 Dwarka, Delhi       0.032258   
104                               Rajouri Garden       0.068966   
106                             Rani Bagh, Delhi       0.066667   
113                        Saket District Centre       0.040000   
112                                Saket (Delhi)       0.040000   
43                        Kabir Nagar, New Delhi       0.074074   
127                                Shivaji Place       0.052632   
116                               Sarojini Nagar       0.028986   

     Cluster Labels   Latitude  Longitude  
137               2  28.646420  77.306150  
134               2  28.531520  77.150200  
96                2  28.614580  77.275740  
5                 2  28.652234  77.129392  
131               2  28.640890  77.089220  
67                2  28.612795  77.288501  
100               2  28.666340  77.125000  
53                2  28.648210  77.142730  
73                2  28.663350  77.140810  
122               2  28.714230  77.157440  
25                2  28.589950  77.040040  
104               2  28.645620  77.122090  
106               2  28.685840  77.131880  
113               2  28.528130  77.219050  
112               2  28.524070  77.206770  
43                2  28.689664  77.141038  
127               2  28.652650  77.121390  
116               2  28.575600  77.193640

### OBSERVATIONS:
Cluster 0 has highest number(116) of shopping mall. On the other hand, cluster 1 has very less shopping malls as compared to other two clusters. This represents a great opportunity and high potential areas to open new shopping malls in cluster 1 as there is very little competition. Meanwhile, shopping malls in cluster 0 are likely suffering from intense competition due to oversupply and high concentration of shopping malls. <br /> 
**Therefore, the outcomes of this project recommend property developers to open new shopping malls in neighborhoods in cluster 1.** <br />
Property developers with unique selling propositions to stand out from the competition can also open new shopping malls in neighborhoods in cluster 2  with moderate competition. <br /> ***Lastly, property developers are advised to avoid neighborhoods in cluster 0 which already have high concentration of shopping malls.***